pkg/apis/serving/v1alpha1/trainedmodel_webhook.go

```diff
@@ -0,0 +1,101 @@
+/*
+Copyright 2020 kubeflow.org.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	"fmt"
+	"github.com/kubeflow/kfserving/pkg/utils"
+	"k8s.io/apimachinery/pkg/runtime"
+	"regexp"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/webhook"
+)
+
+// regular expressions for validation of isvc name
+const (
+	TmNameFmt                   string = "[a-z]([-a-z0-9]*[a-z0-9])?"
+	InvalidTmNameFormatError           = "the Trained Model \"%s\" is invalid: a Trained Model name must consist of lower case alphanumeric characters or '-', and must start with alphabetical character. (e.g. \"my-name\" or \"abc-123\", regex used for validation is '%s')"
+	InvalidTmMemoryModification        = "the Trained Model \"%s\" update is invalid: a Trained Model's Model Spec memory must not be altered after being create. The memory is \"%s\" but the update request memory is \"%s\""
+)
+
+var (
+	// log is for logging in this package.
+	trainedmodellog = logf.Log.WithName("trainedmodel-alpha1-resource")
+	// regular expressions for validation of isvc name
+	TmRegexp = regexp.MustCompile("^" + TmNameFmt + "$")
+)
+
+// +kubebuilder:webhook:verbs=create;update,path=/validate-trainedmodel,mutating=false,failurePolicy=fail,groups=serving.kubeflow.org,resources=trainedmodels,versions=v1alpha1,name=trainedmodel.kfserving-webhook-server.validator
+
+var _ webhook.Validator = &TrainedModel{}
+
+// ValidateCreate implements webhook.Validator so a webhook will be registered for the type
+func (tm *TrainedModel) ValidateCreate() error {
+	trainedmodellog.Info("validate create", "name", tm.Name)
+	// TODO: Validate storageURI
```

Comment:
worth to validate that parent inference service exists + is added to spec

Reply:
Should we move it from the controller to the validation webhook? I was thinking the validation webhook should perform pure schema-based validation whereas the controller performs dependency checks and whatnot.

On another note, by spec do you mean verifying that the name of the inference service in the spec should be valid? Or checking that it is defined?

Reply:
We should have a `InferenceServiceReady` condition in the trained model status

```
  - lastTransitionTime: "2021-02-21T04:41:25Z"
    message: InferenceService sklearn-iris is not ready
    reason: InferenceServiceNotReady
    status: "False"
    type: InferenceServiceReady
```

Reply:
What is the advantage of adding an `InferenceServiceReady`? Couldn't we store the reason in a generic `ready` status if we wanted to know why the trained model is not ready?

Reply:
`Ready` can be the overall condition and it can depend on other conditions `InferenceServiceReady`, this is how works for the InferenceService conditions, the `Ready` condition depends on `PredictorReady`, `TransformerReady` and `ExplainerReady` conditions, if any of the those conditions is false the overall condition `Ready` is then marked false.

Reply:
I think adding an `InferenceServiceReady` status can go into a different PR that validates if the inference service is ready. But yeah sounds great!